clash-lib/src/Clash/Core/Subst.hs

```diff
@@ -503,7 +503,14 @@ lookupIdSubst doc (Subst inScope tmS _ genv) v
                      _      -> Var v
   | Just e <- lookupVarEnv v tmS = e
   -- Vital! See 'IdSubstEnv' Note [Extending the Subst]
-  | Just v' <- lookupInScope inScope v = Var (coerce v')
+  --
+  -- TODO: We match on Id here to workaround an issue where type variables
+  -- TODO: "shadow" term variables. Omitting the check would make 'lookupIdSubst'
+  -- TODO: potentially replace an "Id" with a TyVar. For more information:
+  -- TODO:
+  -- TODO:   https://github.com/clash-lang/clash-compiler/issues/1046
+  -- TODO:
+  | Just v'@(Id {}) <- lookupInScope inScope v = Var (coerce v')
```

Comment:
We shouldn't need the `coerce` anymore, we already know it's a `Var Term` since we pattern mached on `Id {}`

Reply:
It's not a GADT, so I don't think we do.

Reply:
Ah, right

Reply:
Is that an 'approved' then? :)

Reply:
Yes

Reply:
Why isn't `Var` a GADT?

Reply:
So we can use "cheap" run-time `coerce` (it's used in quite a few places in the codebase), instead of "expensive" run-time matching on constructors.

Reply:
The proper solution here is to split up `InScopeSet` into two sets internally, one for term vars and the other for type vars. We wouldn't need coerce or run-time matching on constructors that way.